pkg/scheduler/factory/factory.go

```diff
@@ -544,6 +552,54 @@ func (c *configFactory) invalidatePredicatesForPvcUpdate(old, new *v1.Persistent
 	c.equivalencePodCache.InvalidateCachedPredicateItemOfAllNodes(invalidPredicates)
 }
 
+func (c *configFactory) onStorageClassAdd(obj interface{}) {
+	sc, ok := obj.(*storagev1.StorageClass)
+	if !ok {
+		glog.Errorf("cannot convert to *storagev1.StorageClass: %v", obj)
+		return
+	}
+
+	// Creating a StorageClass with late binding can cause previously errored volume predicates to pass
```

Comment:
Shouldn't you invalidate predicates before moving pods to active queue?

Reply:
Predicate results are cached in both failure and success cases.

Reply:
I think this is slightly different.  IIUC, predicates can return 3 results: success, fail, error.  If it returns error, then the result is [not cached](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/core/generic_scheduler.go#L483).

So in my scenario, the predicate would be returning an error.

Reply:
@msau42 equivalence cache is enabled by default in our integration tests if you use [`initTest`](https://github.com/kubernetes/kubernetes/blob/master/test/integration/scheduler/util.go#L59) to initialize the test environment. This is how most of scheduler's integration tests are initialized.  In order to test invalidation, you should:
1. create a Pod that is, for example, unschedulable.
1. Wait for it to be marked unschedulable.
1. Trigger the invalidation event, for example, by creating a storage class.
1. Wait for the Pod to be scheduled.

By doing this, you ensure that the eCache is invalidated properly.